allensdk/brain_observatory/demixer.py

```diff
@@ -33,23 +33,100 @@
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 #
+from typing import Tuple, Optional
+import os
+import logging
+
+import numpy as np
 import scipy.sparse as sparse
 import scipy.linalg as linalg
-import numpy as np
-import os
 import matplotlib.pyplot as plt
-import logging
 import matplotlib.colors as colors
+
+import allensdk.internal.brain_observatory.mask_set as mask_set
 from allensdk.config.manifest import Manifest
 
-def demix_time_dep_masks(raw_traces, stack, masks):
-    '''
 
-    :param raw_traces: extracted traces
-    :param stack: movie (same length as traces)
-    :param masks: binary roi masks
-    :return: demixed traces
-    '''
+def identify_valid_masks(mask_array):
+    ms = mask_set.MaskSet(masks=mask_array.astype(bool))
+    valid_masks = np.ones(mask_array.shape[0]).astype(bool)
+
+    # detect duplicates
+    duplicates = ms.detect_duplicates(overlap_threshold=0.9)
+    if len(duplicates) > 0:
+        valid_masks[duplicates.keys()] = False
+        
+    # detect unions, only for remaining valid masks
+    valid_idxs = np.where(valid_masks)
+    ms = mask_set.MaskSet(masks=mask_array[valid_idxs].astype(bool))
+    unions = ms.detect_unions()
+
+    if len(unions) > 0:
+        un_idxs = unions.keys()
+        valid_masks[valid_idxs[0][un_idxs]] = False
+
+    return valid_masks
+
+
+def _demix_point(source_frame: np.ndarray, mask_traces: np.ndarray,
+                 flat_masks: sparse,
+                 pixels_per_mask: np.ndarray) -> Optional[np.ndarray]:
+    """
+    Helper function to run demixing for single point in time for a
+    source with overlapping traces.
+
+    Parameters
+    ==========
+    source_frame: values of movie source at the single time point,
+    unraveled in the x-y dimension (1d array of length HxW )
+    flat_masks: 2d-array of binary masks unraveled in the x-y dimension
+    mask traces: values of mask trace at single time point (1d-array of
+        length n, where `n` is number of masks)
+    pixels_per_mask: Number of pixels for each mask associated with
+        trace (1d-array of length `n`)
+
+    Returns
+    =======
+    Array of demixed trace values for each mask if all trace data is
+    nonzero. Otherwise, returns None.
+    """
+    mask_weighted_trace = mask_traces * pixels_per_mask
+
+    # Skip if there is zero signal anywhere in one of the traces
+    if (mask_weighted_trace == 0).any():
```

Comment:
This works, but maybe `np.count_nonzero` would be more straightforward to read?

Reply:
I have to disagree here. I think this is much clearer in intent than 
if `len(mask_weighted_trace) == np.count_nonzero(mask_weighted_trace)`.